File: examples/cpp/consume_operator.cpp
#include <cstddef>
#include <cstdint>
#include <vector>

#include <tsl.hpp>

struct sum_sink {
    std::int64_t total = 0;

    template <class Vec>
    void operator()(typename tsl::reg_param<Vec>::type value) {
        total += static_cast<std::int64_t>(tsl::hadd<Vec>(value));
    }
};

struct pair_sum_sink {
    std::int64_t total = 0;

    template <class Vec>
    void operator()(
        typename tsl::reg_param<Vec>::type left,
        typename tsl::reg_param<Vec>::type right) {
        total += static_cast<std::int64_t>(tsl::hadd<Vec>(tsl::add<Vec>(left, right)));
    }
};

void fill_inputs(
    std::vector<std::int32_t>& left,
    std::vector<std::int32_t>& right) {
    for (std::size_t i = 0; i < left.size(); ++i) {
        left[i] = static_cast<std::int32_t>(static_cast<int>((i * 19) % 67) - 33);
        right[i] = static_cast<std::int32_t>(static_cast<int>((i * 23) % 71) - 35);
    }
}

std::int64_t expected_unary_sum(const std::vector<std::int32_t>& input) {
    std::int64_t total = 0;
    for (const auto value : input) {
        total += value;
    }
    return total;
}

std::int64_t expected_binary_sum(
    const std::vector<std::int32_t>& left,
    const std::vector<std::int32_t>& right) {
    std::int64_t total = 0;
    for (std::size_t i = 0; i < left.size(); ++i) {
        total += static_cast<std::int64_t>(left[i]) + right[i];
    }
    return total;
}

template <std::size_t ParallelN>
bool run_fixed_cases(
    const std::vector<std::int32_t>& left,
    const std::vector<std::int32_t>& right,
    std::int64_t expected_unary,
    std::int64_t expected_binary) {
    sum_sink unary;
    tsl::algo::consume_unary<ParallelN, tsl::algo::alignment::unaligned>(
        unary,
        left.data(),
        left.size());
    if (unary.total != expected_unary) {
        return false;
    }

    pair_sum_sink binary;
    tsl::algo::consume_binary<ParallelN, tsl::algo::alignment::unaligned>(
        binary,
        left.data(),
        right.data(),
        left.size());
    return binary.total == expected_binary;
}

bool run_native_cases(
    const std::vector<std::int32_t>& left,
    const std::vector<std::int32_t>& right,
    std::int64_t expected_unary,
    std::int64_t expected_binary) {
    sum_sink unary;
    tsl::algo::consume_unary<
        tsl::algo::parallelism::native,
        tsl::algo::alignment::unaligned>(
        unary,
        left.data(),
        left.size());
    if (unary.total != expected_unary) {
        return false;
    }

    pair_sum_sink binary;
    tsl::algo::consume_binary<
        tsl::algo::parallelism::native,
        tsl::algo::alignment::unaligned>(
        binary,
        left.data(),
        right.data(),
        left.size());
    return binary.total == expected_binary;
}

int main() {
    constexpr std::size_t count = 1000;
    std::vector<std::int32_t> left(count);
    std::vector<std::int32_t> right(count);
    fill_inputs(left, right);

    const auto expected_unary = expected_unary_sum(left);
    const auto expected_binary = expected_binary_sum(left, right);

    if (!run_native_cases(left, right, expected_unary, expected_binary)) {
        return 1;
    }
    if (!run_fixed_cases<1>(left, right, expected_unary, expected_binary)) {
        return 2;
    }
    if (!run_fixed_cases<4>(left, right, expected_unary, expected_binary)) {
        return 3;
    }
    if (!run_fixed_cases<16>(left, right, expected_unary, expected_binary)) {
        return 4;
    }
    return 0;
}

File: examples/rust/src/bin/consume_operator.rs
use tsl::profile;
use tsl::tsl_core::StaticSimdVector;

struct SumSink {
    total: i64,
}

impl<V> profile::algo::UnaryConsumeKernel<V> for SumSink
where
    V: StaticSimdVector<BaseType = i32> + profile::detail::primitives::HaddImpl,
{
    fn consume(&mut self, value: V::RegisterType) {
        self.total += i64::from(profile::hadd::<V>(value));
    }
}

struct PairSumSink {
    total: i64,
}

impl<V> profile::algo::BinaryConsumeKernel<V> for PairSumSink
where
    V: StaticSimdVector<BaseType = i32>
        + profile::detail::primitives::AddImpl
        + profile::detail::primitives::HaddImpl,
{
    fn consume(&mut self, left: V::RegisterType, right: V::RegisterType) {
        let sum = profile::add::<V>(left, right);
        self.total += i64::from(profile::hadd::<V>(sum));
    }
}

fn fill_inputs(left: &mut [i32], right: &mut [i32]) {
    for (i, (left_value, right_value)) in left.iter_mut().zip(right.iter_mut()).enumerate() {
        *left_value = ((i * 19) % 67) as i32 - 33;
        *right_value = ((i * 23) % 71) as i32 - 35;
    }
}

fn expected_unary_sum(input: &[i32]) -> i64 {
    input.iter().map(|value| i64::from(*value)).sum()
}

fn expected_binary_sum(left: &[i32], right: &[i32]) -> i64 {
    left.iter()
        .zip(right.iter())
        .map(|(left_value, right_value)| i64::from(*left_value) + i64::from(*right_value))
        .sum()
}

fn main() {
    let mut left = vec![0i32; 1000];
    let mut right = vec![0i32; 1000];
    fill_inputs(&mut left, &mut right);

    let expected_unary = expected_unary_sum(&left);
    let expected_binary = expected_binary_sum(&left, &right);

    macro_rules! run_policy {
        ($policy:expr) => {{
            let mut unary = SumSink { total: 0 };
            profile::algo::consume_unary($policy, &mut unary, &left);
            assert_eq!(unary.total, expected_unary);

            let mut binary = PairSumSink { total: 0 };
            profile::algo::consume_binary($policy, &mut binary, &left, &right);
            assert_eq!(binary.total, expected_binary);
        }};
    }

    run_policy!(tsl::dataparallel::fixed::<1>());
    run_policy!(tsl::dataparallel::generic::<4>());
    run_policy!(tsl::dataparallel::generic::<16>());
}
